File: Windows/BotSpeaker/OrchestratedMeetingTemplate.cs
using System.Text.RegularExpressions;

namespace BotSpeaker;

public sealed record OrchestratedScriptTurn(int SpeakerIndex, string Text);

public sealed class OrchestratedSpeakerConfiguration
{
    public required int Slot { get; init; }
    public required string Role { get; init; }
    public string Name { get; set; } = "";
    public string VoiceId { get; set; } = "";
    public string VoiceName { get; set; } = "";
    public string Placeholder => $"{{{{speaker_{Slot}}}}}";
}

public sealed record OrchestratedMeetingTemplate(
    string Id,
    string Title,
    string Detail,
    IReadOnlyList<string> SpeakerRoles,
    IReadOnlyList<string> DefaultVoiceGenders,
    string Text)
{
    public int SpeakerCount => SpeakerRoles.Count;
    public int TurnCount => ParseTurns(Text).Count;

    public List<OrchestratedScriptTurn> ParseTurns(string text)
    {
        var paragraphs = Regex.Split(text.Replace("\r\n", "\n").Trim(), @"\n\s*\n")
            .Select(value => value.Trim())
            .Where(value => value.Length > 0)
            .ToList();
        var result = new List<OrchestratedScriptTurn>();
        for (int index = 0; index < paragraphs.Count; index++)
        {
            var match = Regex.Match(
                paragraphs[index],
                @"^\{\{speaker_(\d+)\}\}\s*:\s*(.+)$",
                RegexOptions.Singleline);
            if (!match.Success
                || !int.TryParse(match.Groups[1].Value, out int oneBasedSpeaker)
                || oneBasedSpeaker < 1
                || oneBasedSpeaker > SpeakerCount)
            {
                throw new AppException(
                    $"Paragraph {index + 1} must start with {{{{speaker_1}}}} through " +
                    $"{{{{speaker_{SpeakerCount}}}}} followed by a colon.");
            }
            result.Add(new OrchestratedScriptTurn(oneBasedSpeaker - 1, match.Groups[2].Value.Trim()));
        }
        if (result.Count == 0) throw new AppException("The orchestrated meeting script is empty.");
        return result;
    }

    public static readonly OrchestratedMeetingTemplate LaunchReadiness = new(
        "meeting-assistant-launch-readiness",
        "AI meeting assistant launch readiness",
        "4 speakers · expressive launch decision",
        ["Product Manager", "Engineering Lead", "Privacy & Security Lead", "Customer Success Lead"],
        ["male", "female", "male", "female"],
        """
        {{speaker_1}}: Alright, everyone's here — let's do this. I'm {{speaker_1}}, product manager for Decision Digest, and today we decide whether this thing actually ships. [exhales] I've rescheduled this meeting twice, so I'm not leaving without an answer. Quick reality check: the launch is five design partners, internal meetings only, post-meeting summaries with every line linked back to the transcript. Nothing auto-sends anywhere. That's the whole product. Where are we?

        {{speaker_2}}: I'm {{speaker_2}}, engineering lead, and the honest answer is "conditional go," which I realize is the least satisfying phrase in software. [laughs] The service is stable at pilot volume and the kill switch works — I tested it myself on Friday, and yes, it actually kills things. Two gates left: export events still don't show up reliably in the audit log, and the batching change needs forty-eight hours of clean canary. That's it. But I want those actually done, not "the PR is basically approved" done.

        {{speaker_3}}: I'm {{speaker_3}}, privacy and security. Conditional from me too, and my condition is NOT negotiable. [firm] If an attendee can't tell that processing is happening, we don't generate. Full stop. I also want one end-to-end deletion drill — a real participant request, all the way through to verified removal of the digest and the embeddings. Not a database command. A request.

        {{speaker_4}}: I'm {{speaker_4}}, customer success, and I bring good news for once! [cheerful] Customers love this thing. Pilot teams say it saves them ten, fifteen minutes per meeting. The bad news — [sighs] — is they keep calling the draft "the record." Which it is not. If we don't hammer "editable draft" into every screen and every training doc, someone is going to file an action item straight from a hallucination, and the escalation will have my name on it.

        {{speaker_1}}: [laughs] Noted. Okay, latency. Median is six seconds, but the tail reaches the high twenties. Do we need a week of shadowing, or is forty-eight hours enough? I want a number, not a vibe. And before anyone says "it depends"—

        {{speaker_2}}: [jumping in] —It depends. [laughs] Kidding! Forty-eight hours, if we define the evidence up front. Replay on ten thousand meetings already brought the ninety-fifth percentile under eleven seconds with no quality drop. Shadow traffic, one canary org, a rollback exercise, and if anything crosses a threshold, the clock resets. Adding five more days after clean evidence is just… superstition. [dry] Comforting, expensive superstition.

        {{speaker_3}}: Fine — but batching cannot cross region boundaries. If a latency win leaks a transcript out of region, I will personally pull the plug, and I will not be graceful about it. [half-joking]

        {{speaker_2}}: It can't. Batches form inside the org-and-region boundary, and there's an alert on any routing mismatch. You'd get paged before I do, which honestly seems fair.

        {{speaker_4}}: One operational thing — when the kill switch fires, the meetings stuck in the queue can't just say "processing" forever. Customers would rather see "delayed by your admin" than a spinner of lies.

        {{speaker_2}}: Agreed. There's a real "delayed" state, and support gets the exact state table before training. Screenshots and all.

        {{speaker_1}}: Okay, German. [sighs] Action-item recall is still a few points behind English. I say we keep it, label it beta, and gate it per language. Objections?

        {{speaker_4}}: Keep it — two of the five customers joined specifically for German. But the account team says "beta" out loud before enablement, and the weekly report breaks omissions out by language. No hiding it in a blended average.

        {{speaker_3}}: Agreed, as long as a German regression pauses German only. And on my deletion drill: support runs it while engineering watches. If the runbook needs an engineer at the keyboard, it failed. My no-go lines are written down — notice failure entering generation, derived content outliving its window, or transcript text showing up in audit logs.

        {{speaker_1}}: Then here's the decision: conditional go, five customers. {{speaker_2}} owns audit events and the canary. {{speaker_3}} owns notice and the deletion drill. {{speaker_4}} owns training and the escalation roster. Any function can pull the brake without a committee vote. [relieved] And can I just say — third attempt at scheduling this meeting, and we actually made a decision. I'm framing this calendar invite. Thanks, everyone.
        """);

    public static readonly OrchestratedMeetingTemplate ApiIncidentReview = new(
        "api-latency-incident-review",
        "API latency incident review",
        "3 speakers · expressive retrospective",
        ["Incident Commander", "Site Reliability Engineer", "Customer Support Lead"],
        ["male", "female", "male"],
        """
        {{speaker_1}}: I'm {{speaker_1}}, incident commander for yesterday's API latency mess. Blameless review, so nobody's getting fired — [wry] — though the config system is on thin ice. Timeline: nine forty-two a.m., checkout latency blows through the two-second objective. Mitigated at ten thirty-one, recovered at ten forty-seven. {{speaker_2}}, walk us through it.

        {{speaker_2}}: I'm {{speaker_2}}, the on-call SRE, running on four hours of sleep. [tired laugh] Nine thirty-eight: a routine config change raised the pricing service's connection pool limit. Fine in staging. In production, every instance opened connections at once and the database hit lock contention. And here's the fun part — [sarcastic] — every alert stayed GREEN. Meanwhile customers couldn't buy ANYTHING.

        {{speaker_3}}: I'm {{speaker_3}}, customer support lead, and customers noticed before our alerts did, which — [sighs] — is becoming a theme. First chat at nine forty-three: a retailer watching a checkout spinner for a full minute. By nine forty-seven, six reports across three queues — nobody saw the pattern until an agent posted them in one channel.

        {{speaker_1}}: Once the page finally fired, what made diagnosis so slow?

        {{speaker_2}}: [frustrated] The rollout dashboard said one hundred percent success — because technically the config did deploy successfully. [deadpan] Successfully broke everything. I wasted twenty minutes chasing the payment provider — slow-looking spans that were just queued behind our own thread pool. At ten oh-two I lined traces up against the deploy timeline — one relevant change all morning.

        {{speaker_3}}: Meanwhile agents called it "intermittent slowness" because some checkouts eventually finished. A forty-second checkout is not slow — it's a customer closing the tab. Severity should follow the user's task, not the HTTP status.

        {{speaker_1}}: Agreed. A two hundred that takes forty seconds is an outage wearing a disguise. So when did we finally—

        {{speaker_2}}: [jumping in] —Roll back? Ten oh-eight, after I reproduced the lock wait on a read replica. And then — [exasperated sigh] — seven minutes waiting for approval: the config owner wasn't in the channel and the runbook listed a team reorged away last year. Rollback ran ten fifteen to ten twenty-two. Latency dropped immediately.

        {{speaker_1}}: Seven minutes of a sev-one waiting for permission — that one's on me. Effective today, incident commanders can authorize rollback of reversible changes. {{speaker_3}}, communications?

        {{speaker_3}}: The status post went up at ten twelve, twenty-nine minutes in, saying we were "investigating elevated latency" — which, honestly, describes every day of my life. [dry] Customers wanted one answer: will retrying double-charge me? We couldn't say no until ten twenty. I want pre-approved language for delayed transactions and a dashboard my agents can actually read.

        {{speaker_2}}: The idempotency keys held — zero duplicates — but support shouldn't page me to learn that; I'll add a panel. Plus two fixes: a five-minute burn-rate alert on checkout latency — it would have fired at nine forty-one, before the first customer chat — and a ten-percent canary with an automatic halt.

        {{speaker_3}}: And I want ticket clustering — three enterprise customers, same symptom, ten minutes, flag it — even when filed under different categories. Plus a follow-up note to the forty-two affected accounts. Just those. Not a company-wide apology blast.

        {{speaker_1}}: Done. Owners: {{speaker_2}} — canary guard, lock-wait dashboard, burn alert, by Friday. Me — rollback authority and ownership audit by Wednesday. {{speaker_3}} — clustering and customer follow-up by Monday. The big lesson: measure availability at the checkout button, not the CPU graph. [warm] Nice recovery, both of you. Review goes out today.
        """);

    public static readonly OrchestratedMeetingTemplate PerformanceReviewOneOnOne = new(
        "performance-review-one-on-one",
        "Performance review 1:1",
        "2 speakers · candid feedback and growth",
        ["Engineering Manager", "Senior Software Engineer"],
        ["female", "male"],
        """
        {{speaker_1}}: Come in, sit down — and {{speaker_2}}, you can stop looking like you're walking into a root canal. [laughs] This is a good review. I'm {{speaker_1}}, your engineering manager, for the record. Short version first, because I know you hate suspense: strong year, exceeding expectations. The billing platform is more reliable because of you, and half the team asks "what would {{speaker_2}} do" during incidents, which is either a compliment or a cry for help.

        {{speaker_2}}: [relieved sigh] Okay. Wow. I'm {{speaker_2}}, senior engineer on billing — and yes, I was braced for the "growth areas" speech. Exceeding expectations feels… really good to hear, honestly. Especially since the ledger migration ran three weeks over and I've been quietly stewing about that since March.

        {{speaker_1}}: The migration going long is not the story. The story is you caught the retry bug that would have double-charged customers before it shipped. Reconciliation errors down sixty percent, and support got hours of their week back. [whispers] Nobody remembers a schedule slip. Everybody remembers double charges. The actual reservation from calibration was visibility — people outside our team sometimes learned about risks later than they needed to.

        {{speaker_2}}: [sighs] Yeah. That one's fair, and honestly I've heard it before. I hold updates until I actually understand the problem, because "something's wrong, more soon" feels like noise. During the database saturation thing I sat on the first update for an hour while I worked out whether it was traffic or the query plan… and meanwhile payments kept deploying straight into the blast radius. Point taken.

        {{speaker_1}}: It's not about broadcasting every theory. Four lines: what we know, what we don't, who's affected, when the next update comes. Your diagnosis was excellent — the silence around it was the problem. Now go ahead and ask the question you've been sitting on since you—

        {{speaker_2}}: [jumping in] Why not promotion? [laughs] Sorry. You clearly saw that coming. I led a cross-team migration, mentored two people through promo packets, and covered a staff gap for six months. I'm not angry, but… [frustrated] okay, I'm a little frustrated. What was missing?

        {{speaker_1}}: Completely fair question, and you deserve a real answer, not committee-speak. The gap is influence before execution. You aligned the database and billing teams after the design was mature — analytics and support found out late, and it cost us rework. At staff level you shape that stakeholder mess early and deliver through other people. Nobody doubts your judgment or your code. The question is whether the organization sees you leading beyond it.

        {{speaker_2}}: [reluctantly] That… lands. I kept pulling implementation back to myself because the schedule felt fragile, and handing someone an unclear problem felt like a trap. Which protected the launch and torpedoed the promo evidence at the same time. If the next project's definition of success is me personally closing the hardest tickets again, we've learned nothing.

        {{speaker_1}}: Which is exactly why I'm giving you account-event consolidation. Four teams, messy ownership, starts next month. You set the technical strategy, run the decision forum, and delegate the designs. I'll get the directors to sign a charter so you're not leading on vibes alone. [wry] And if I catch you writing half the implementation at two a.m., we're having a very different meeting.

        {{speaker_2}}: [laughs] Deal. But let's define success now: four teams aligned before implementation starts, each workstream with an owner who can defend the design without me in the room, and a weekly decision note people actually read. And midpoint feedback — I'm not waiting a year to find out I did it wrong again.

        {{speaker_1}}: Agreed on all of it. On money: the raise reflects the rating, final numbers next week. On promotion: show staff-level leverage through the midpoint and I'll take your case to the midyear window in May. I can't promise the outcome, but I promise you won't be guessing. And {{speaker_2}} — [warm] — genuinely well done this year. Now get out of my office before I start listing growth areas. [laughs]
        """);

    public static readonly IReadOnlyList<OrchestratedMeetingTemplate> All =
        [LaunchReadiness, ApiIncidentReview, PerformanceReviewOneOnOne];
}

File: macOS/BotSpeaker/OrchestratedMeetingTemplate.swift
import Foundation

struct OrchestratedMeetingTemplate: Identifiable, Hashable {
    let id: String
    let title: String
    let detail: String
    let speakerRoles: [String]
    let defaultVoiceGenders: [String]
    let text: String

    var speakerCount: Int { speakerRoles.count }
    var turnCount: Int { (try? parsedTurns().count) ?? 0 }

    func parsedTurns() throws -> [OrchestratedScriptTurn] {
        try OrchestratedScriptParser.parse(text, speakerCount: speakerCount)
    }

    static let launchReadiness = OrchestratedMeetingTemplate(
        id: "meeting-assistant-launch-readiness",
        title: "AI meeting assistant launch readiness",
        detail: "4 speakers · natural launch decision",
        speakerRoles: ["Product Manager", "Engineering Lead", "Privacy & Security Lead", "Customer Success Lead"],
        defaultVoiceGenders: ["male", "female", "male", "female"],
        text: launchReadinessText
    )

    static let apiIncidentReview = OrchestratedMeetingTemplate(
        id: "api-latency-incident-review",
        title: "API latency incident review",
        detail: "3 speakers · operational retrospective",
        speakerRoles: ["Incident Commander", "Site Reliability Engineer", "Customer Support Lead"],
        defaultVoiceGenders: ["male", "female", "male"],
        text: apiIncidentReviewText
    )

    static let all = [launchReadiness, apiIncidentReview]

    private static let apiIncidentReviewText = """
    {{speaker_1}}: I am {{speaker_1}}, the Incident Commander for yesterday’s API latency event. This is a learning review, not a search for individual fault. We will reconstruct what customers experienced, identify where our detection and coordination helped or slowed us down, and leave with specific owners. The incident began at 9:42 a.m. Pacific when checkout requests in the western region started exceeding our two-second objective. We mitigated the customer impact at 10:31 and declared recovery at 10:47 after the queues drained. {{speaker_2}}, please start with the technical sequence.

    {{speaker_2}}: I am {{speaker_2}}, the Site Reliability Engineer who was primary on call. At 9:38, a routine configuration rollout increased the connection pool limit for the pricing service. The new limit looked safe in staging, but production traffic caused every application instance to open connections at nearly the same time. The database stayed available, yet lock contention increased sharply. Pricing calls slowed, checkout workers accumulated, and retries amplified the load. Our infrastructure alerts stayed green because CPU, memory, and error rate were normal. The first page came from the checkout latency objective at 9:46.

    {{speaker_3}}: I am {{speaker_3}}, the Customer Support Lead. Customers noticed before that page. The first chat arrived at 9:43 from a retailer whose checkout spinner ran for almost a minute. By 9:47 we had six similar contacts, but the tickets were split between payments, storefront, and general performance queues. The common pattern became obvious only when an agent posted the examples in the incident channel at 9:52. We should discuss why that customer signal did not reach the on-call team earlier.

    {{speaker_1}}: Agreed. First, on diagnosis: once the latency alert fired, what made the connection change difficult to identify?

    {{speaker_2}}: Two things. The rollout dashboard showed one hundred percent success because the configuration reached every instance and passed its health check. Also, our database dashboard emphasizes utilization and failed connections, not concurrent lock wait. I initially investigated the payment provider because its span was the slowest child in several traces. That was misleading: the provider span included time waiting for a worker thread. At 10:02 I compared a slow trace with the deployment timeline, and the pricing configuration was the only relevant change.

    {{speaker_3}}: That matches what Support saw. Some checkouts eventually completed, so agents hesitated to call it an outage. They described it as intermittent slowness, while customers described it as being unable to buy. We need severity guidance based on the user’s task, not only on HTTP failure rate.

    {{speaker_1}}: Yes. A checkout that takes forty seconds is functionally unavailable even if it returns status two hundred. When did we decide to roll back?

    {{speaker_2}}: At 10:08. I proposed the rollback after reproducing the lock wait in a production read replica. Approval took seven minutes because the configuration owner was not in the incident channel and our runbook still listed the previous team. The rollback began at 10:15, completed at 10:22, and latency improved immediately. We then reduced retry concurrency so the accumulated queue could drain without causing another spike.

    {{speaker_1}}: That approval delay is actionable. For a reversible configuration tied to an active severity-one symptom, the incident commander should be able to authorize rollback. We can notify the owning team in parallel. {{speaker_3}}, how did communication go during that window?

    {{speaker_3}}: The internal update was useful, but the public status message came too late. We posted at 10:12, twenty-nine minutes after the first customer report. The message said we were investigating elevated API latency, which was accurate but too broad. Checkout customers wanted to know whether retrying would create duplicate orders. Agents did not have an approved answer until 10:20. We should prepare symptom-specific language for delayed transactions and state clearly when retries are safe.

    {{speaker_2}}: For this event, duplicate orders were prevented by idempotency keys, but Support could not verify that from the dashboard. I can add an incident panel showing duplicate-suppression activity and queue depth. That would let the communication lead answer with evidence rather than waiting for an engineer.

    {{speaker_3}}: That would help. I also want ticket clustering to alert us when three enterprise customers report the same workflow symptom within ten minutes, even if they choose different categories. It should suggest a pattern to the duty lead, not automatically declare an incident.

    {{speaker_1}}: Good distinction. Let’s cover detection. {{speaker_2}}, what alert would have fired closest to the actual beginning without becoming noisy during normal peaks?

    {{speaker_2}}: A burn-rate alert on checkout duration segmented by region would have fired around 9:41. We already calculate that service-level indicator, but paging uses a fifteen-minute window. I propose a fast five-minute burn alert paired with queue growth, plus a dashboard for database lock-wait time. I do not recommend paging directly on connection count because healthy traffic bursts can produce the same shape.

    {{speaker_3}}: Can the alert include the affected customer workflow in plain language? “Checkout completion delayed in us-west” helps Support far more than a service identifier.

    {{speaker_2}}: Yes. The alert metadata can map the dependency to checkout and link the customer-impact dashboard. I’ll include that in the same change.

    {{speaker_1}}: Now prevention. The configuration itself was reviewed, so what test or rollout control was missing?

    {{speaker_2}}: Staging has too few application instances to reproduce the connection fan-out. We should validate aggregate connection demand as part of configuration review, cap the rollout to ten percent of production instances, and hold for ten minutes while comparing lock wait and checkout latency. The deployment controller can automatically halt if either metric regresses. I also want the connection pool limit expressed as a regional budget divided among instances instead of a fixed per-instance number.

    {{speaker_3}}: From the customer side, I want a short follow-up for affected accounts that explains the symptom, confirms that completed orders were not duplicated, and names the prevention work without exposing internal details. We should send it to the forty-two accounts that crossed the delay threshold, not to every customer.

    {{speaker_1}}: Agreed. Let me summarize the actions. {{speaker_2}} owns the canary rollout guard, regional connection budget, lock-wait dashboard, and faster checkout burn alert by next Friday. I own the rollback authority update and the service ownership audit by Wednesday. {{speaker_3}} owns the ticket-pattern proposal, support guidance for delayed transactions, and targeted customer follow-up by Monday. We will test the revised runbook in a game day within two weeks.

    {{speaker_2}}: That captures my items. I’ll also attach the query and trace evidence from this review so the alert thresholds can be evaluated against the actual incident.

    {{speaker_3}}: Mine too. I’ll include the frontline agents who handled the first contacts in the communication review; they saw the ambiguity before the rest of us did.

    {{speaker_1}}: Perfect. The main lesson is that availability must be measured at the customer workflow, not inferred from healthy infrastructure. We recovered safely, but detection, rollback authority, and customer guidance were slower than they should have been. Thank you both. I’ll publish the review with these owners and dates today.
    """

    private static let launchReadinessText = """
    {{speaker_1}}: I am {{speaker_1}}, the Product Manager for Decision Digest. Thanks for making the time. The purpose of this review is to decide whether our AI meeting assistant is ready for a controlled launch, not to defend a date we already picked. I want a direct answer on product value, reliability, privacy, and customer operations. If the answer is conditional, we will name the conditions, owners, and stop signals before we leave. For context, the first cohort is five design partners using post-meeting summaries for scheduled internal meetings. The assistant drafts decisions, action items, and open questions, and every item links back to transcript evidence. Nothing is sent automatically to task systems or external recipients.

    {{speaker_2}}: I am {{speaker_2}}, the Engineering Lead. That scope matches what we have tested. The service is stable at pilot volume, and the architecture gives us a clean kill switch before a transcript enters generation. My current recommendation is a conditional go. Two technical gates remain: the export event must appear reliably in the workspace audit log, and the batching optimization must finish its live canary without moving the quality metrics. Those are bounded pieces of work, but I do not want us to blur “nearly merged” with “production evidence.”

    {{speaker_3}}: I am {{speaker_3}}, the Privacy and Security Lead. I agree with the conditional framing. My question is slightly different: can an attendee understand that processing is happening, and can we honor a withdrawal without asking support to improvise? Administrator enablement is necessary, but it is not enough for a live conversation. We require an organizer action, persistent attendee notice, a recorded notice-delivery event, and a hard rejection when notice delivery fails. I also need one end-to-end deletion exercise that starts with a participant request and ends with verified removal of the derived digest, embeddings, and temporary inference artifacts.

    {{speaker_4}}: I am {{speaker_4}}, the Customer Success Lead. From the customer side, there is real pull. Pilot teams repeatedly tell us they recover ten to fifteen minutes after a planning meeting, especially when action items are scattered across chat and personal notes. But they also call the draft “the record” unless we teach them otherwise. I support the five-customer cohort if the product and launch materials consistently call it an editable, transcript-linked draft. Support needs a capability matrix, an escalation path, and a clear answer for an attendee who says, “I did not want this summary created.”

    {{speaker_1}}: Good. Let me test the value proposition against those boundaries. We are promising faster follow-through with evidence, not perfect minutes and not an autonomous source of truth. The organizer reviews before sharing. External meetings, webinars, coaching scores, regulated templates, and automatic task creation remain out of scope. Are we comfortable that this is still useful enough for the cohort, or have the safeguards removed the reason customers wanted it?

    {{speaker_4}}: It is still useful. Actually, the review step is part of the appeal for the administrators I spoke with. They do not want another bot silently publishing tasks. They want a strong first draft and a quick way to verify attribution. One caveat: the empty-state copy should say why a digest was blocked when notice validation fails. If it just says “generation unavailable,” administrators will open tickets and may retry in ways that confuse attendees.

    {{speaker_2}}: Yep, that is straightforward. We already expose a structured rejection reason internally. We can map the notice-validation case to customer-safe copy without exposing infrastructure details. I will add that to the launch branch and include it in the end-to-end test.

    {{speaker_3}}: That works for me, with one wording review. Please avoid implying the attendee caused the failure. Something like, “The digest was not created because attendee notice could not be confirmed.” It is factual and it reinforces the safeguard.

    {{speaker_1}}: Agreed. Now latency. Our pilot median is about six seconds after the meeting ends, the ninety-fifth percentile is just under fifteen, and the long tail reaches the high twenties. The launch promise says the draft appears shortly after the call. Do we need a full week of production shadowing, or is forty-eight hours enough?

    {{speaker_2}}: Forty-eight hours is enough if we define the evidence before the clock starts. The batching change groups compatible transcript windows and removes queue overhead. Replay on ten thousand pilot meetings brought the ninety-fifth percentile below eleven seconds with no measurable drop in grounding or action-item precision. I want forty-eight hours of shadow traffic, one canary organization, language-segmented quality, and a rollback exercise. If any comparison crosses the threshold, the clock resets. Waiting seven days after clean evidence would add calendar time, not confidence.

    {{speaker_3}}: I’m comfortable with that. Is regional isolation included in the comparison? A latency win is not acceptable if a fallback can send content outside the workspace’s configured region.

    {{speaker_2}}: Yes. Batches are formed only inside the organization and region boundary. Admission queues work rather than borrowing capacity from another region. We have an alert on any attempted routing mismatch, and the canary dashboard breaks that out separately.

    {{speaker_3}}: Great. Then no objection from me on forty-eight hours.

    {{speaker_4}}: One operational question: when the kill switch stops new jobs, what does a customer see for meetings already waiting? “Processing” forever would be worse than a clear delay state.

    {{speaker_2}}: In-flight jobs either finish on the pinned version or move to “delayed by administrator control,” depending on where they are in the pipeline. We do not discard the source meeting. Once the hold is lifted, eligible jobs resume. I can give Support the exact state table and screenshots before training.

    {{speaker_4}}: Perfect. That gives us something concrete to practice.

    {{speaker_1}}: Let’s cover language scope. English, Spanish, and French meet the agreed thresholds. German action-item recall is still several points below English, although precision remains high. My proposal is to keep all four languages in the cohort, label non-English output as beta, and make the readiness dashboard language-specific. Thoughts?

    {{speaker_4}}: I would keep German for two of the five customers because they explicitly joined to test it. Removing it now loses useful evidence. But the account team must say “beta” before enablement, and the weekly report must show omissions by language rather than one blended satisfaction score.

    {{speaker_3}}: I agree, provided the label is visible where the digest is reviewed, not buried in an administrator document. Also, a language regression should pause that language independently. We should not disable healthy English processing because German recall drops, and we should not hide German risk inside a healthy global average.

    {{speaker_2}}: We can gate by language. The classifier change recovers four points in replay, and the deployment system already supports a language allowlist. I will add per-language stop thresholds to the launch configuration.

    {{speaker_1}}: Sounds aligned. Privacy, walk us through the withdrawal test and what would make you say no-go.

    {{speaker_3}}: The test begins as a real request, not a database command. Customer Success submits the request through the documented channel with a meeting identifier and participant context. We verify authorization, delete the entire derived digest, remove processing artifacts, and record acknowledgments from every subsystem. The source recording follows the customer’s separate retention policy, so the confirmation must not claim we deleted that too. I will say no-go if notice failure can still enter generation, if derived content survives beyond the documented window, if support needs privileged engineering access to complete the request, or if audit evidence contains transcript excerpts rather than identifiers and event types.

    {{speaker_4}}: I’d like Support to run the request while Engineering observes. That proves the procedure works for the team who will actually receive it.

    {{speaker_2}}: Agreed. We will instrument the exercise, but we won’t take over the steps. If the runbook cannot be followed without an engineer at the keyboard, it has not passed.

    {{speaker_1}}: Excellent. Let’s turn to audit logging. Today an organizer can copy or export a digest, but the workspace audit log does not reliably record every export path. I consider that a hard launch gate because administrators use the log to investigate where meeting content went.

    {{speaker_3}}: Strong agreement. Viewing and exporting are different risk events. We need actor, workspace, digest identifier, export type, and timestamp. No transcript text in the event. Failed export attempts should be distinguishable from completed exports, and the retention must match the customer’s audit policy.

    {{speaker_2}}: The event is implemented for download and copy. The remaining path is the share-sheet handoff. It will be in Thursday’s build. We can test all three paths against an isolated workspace and attach the resulting audit records to the launch checklist.

    {{speaker_4}}: That also makes support conversations easier. When an administrator asks whether a digest left the product, we can point them to evidence instead of guessing from application logs.

    {{speaker_1}}: All right. I’m hearing four launch conditions: complete export auditing, pass the forty-eight-hour latency canary and rollback exercise, pass notice and withdrawal tests end to end, and finish customer-facing training with the exact states and claims. Is anything missing?

    {{speaker_2}}: Capacity reservation. It is already arranged, but it belongs in the signed record. The first cohort has reserved model-provider capacity and thirty percent headroom at the expected peak. If quota falls below that buffer, admission slows before quality or isolation is compromised.

    {{speaker_4}}: Add an escalation roster with one named engineering lead and one Customer Success owner each business day. A dashboard without a person watching it is not a launch control.

    {{speaker_3}}: And each function needs authority to trigger a hold. The launch commander makes the final coordination call, but Engineering, Privacy, Security, or Customer Success should not need consensus to stop expansion when its threshold is crossed.

    {{speaker_1}}: Yes to all three. Here is the decision I’ll record: conditional go for five customers, not unrestricted availability. {{speaker_2}} owns export auditing, canary evidence, capacity confirmation, and rollback. {{speaker_3}} owns notice and deletion verification plus the privacy wording. {{speaker_4}} owns cohort checklists, training, and the escalation roster. I own the capability statement, launch record, and daily review. Expansion pauses automatically on any stop threshold, and we reconvene after five hundred production digests.

    {{speaker_4}}: That is clear. I’ll have the customer checklist and simulated support cases ready by Monday.

    {{speaker_3}}: Conditional approval from me, assuming the evidence is attached rather than summarized as a checkbox.

    {{speaker_2}}: Same here. I’ll publish the canary dashboard and rollback timeline so everyone can review the raw result.

    {{speaker_1}}: Great. Then we have a decision. We are launching a reviewable assistant that reduces administrative work while keeping evidence and participant control visible. We are not launching an autonomous record keeper. I’ll send the decision log today with owners, dates, and stop conditions. Thanks, everyone.
    """
}

struct OrchestratedScriptTurn: Hashable {
    let speakerIndex: Int
    let text: String
}

struct OrchestratedSpeakerConfiguration: Identifiable, Hashable {
    let slot: Int
    let role: String
    var name: String
    var voiceID: String
    var voiceName: String

    var id: Int { slot }
    var placeholder: String { "{{speaker_\(slot)}}" }
}

enum OrchestratedScriptParser {
    static func parse(_ text: String, speakerCount: Int) throws -> [OrchestratedScriptTurn] {
        let pattern = #"^\{\{speaker_(\d+)\}\}\s*:\s*(.+)$"#
        let expression = try NSRegularExpression(pattern: pattern, options: [.dotMatchesLineSeparators])
        let paragraphs = text.replacingOccurrences(of: "\r\n", with: "\n")
            .components(separatedBy: "\n\n")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        var turns: [OrchestratedScriptTurn] = []
        for (offset, paragraph) in paragraphs.enumerated() {
            let range = NSRange(paragraph.startIndex..., in: paragraph)
            guard let match = expression.firstMatch(in: paragraph, range: range),
                  match.range == range,
                  let speakerRange = Range(match.range(at: 1), in: paragraph),
                  let bodyRange = Range(match.range(at: 2), in: paragraph),
                  let oneBasedIndex = Int(paragraph[speakerRange]),
                  (1...speakerCount).contains(oneBasedIndex) else {
                throw AppError("Paragraph \(offset + 1) must start with {{speaker_1}} through {{speaker_\(speakerCount)}} followed by a colon.")
            }
            turns.append(OrchestratedScriptTurn(
                speakerIndex: oneBasedIndex - 1,
                text: String(paragraph[bodyRange]).trimmingCharacters(in: .whitespacesAndNewlines)
            ))
        }
        guard !turns.isEmpty else { throw AppError("The orchestrated meeting script is empty.") }
        return turns
    }
}
